### ISSUE TITLE
Output parametric SCAD files

### ISSUE THREAD

**Issue by sparr** [jeff-dh/SolidPython#66]
It's nice to be able to use python for control structures and other language features. However, once the SCAD file has been generated, I'd like to let other people (including the Thingiverse Customizer platform) change parameters of the model.

I'd like for the generated SCAD file to use SCAD variables and operations wherever possible. I know this would be very difficult/complex to implement, so I'm not expecting it to get implemented any time soon, but am hopeful that this issue will prompt some discussion about whether it's worthwhile, how it might work, pointers toward implementation, etc.

**Comment by jeff-dh** [comment]
It's already implemented:

https://github.com/jeff-dh/SolidPython/blob/master-2.0.0-beta-dev/solid2/examples/10-customizer.py

**Comment by sparr** [comment]
Oh, thanks, that looks like it would do most of what I was expecting.

I'd still hope for being able to do at least a loop, possibly more pythonic control structures, without having to scad_inline the whole thing.

This code replacing the scad_inline'd loop gets a little closer, allowing for solidpython / python code inside the loop instead of just literal scad code, but is unwieldy / dangerous and produces not ideal indenting in the scad code:

```python
i = ScadVariable("i", 1)
scene += scad_inline("for (i = [1:objects]){")
scene += cube(side).translateX(2*i*side)
scene += scad_inline("}")
```

PS: is `union()` from that example the convention/standard/appropriate way to create a blank scene?

**Comment by jeff-dh** [comment]
I use union as an empty container. Solidpython(1) used to have part, which was removed and union serves as replacement.

Concerning the loop issue: it's technically not possible. Think about the different execution times. Python gets executed and produces the scad file. The scad file gets customized in openscad afterwards! There's no way for solidpython to access the customized variables....

05.10.2024 13:43:03 sparr ***@***.***>:

**Comment by sparr** [comment]
The example in my code shows that it is possible to do a SCAD loop that contains python-generated code. My approach is unsafe and unwieldy, but it does work.

I'm imagining an approach that might look something like this:

```python
i = ScadVariable("i", 1)
scene += scad_for(i, range(1, objects+1), lambda:
    return cube(side).translateX(2*i*side)
)
```

This would require ~two new features:

1. `scad_for` which would emit the loop header, then run the function in its third parameter, then emit the trailing `}` of the loop, and the whole thing would return a ScadObject(?)
2. SCAD serialization of `range` objects.

Does that seem viable?

**Comment by jeff-dh** [comment]
objects is your customized parameter? How do you want to "translate"
the objects object to the coresponding scad variable?

**Comment by jeff-dh** [comment]
if you have any proposals or solutions feel free to go for it.

Either write an extension (if it's a couple of functions that can be
imported) or extend solidpython an propose a PR.

I could support you, but I don't think I have the time to take a
look at it myself atm.

**Comment by sparr** [comment]
Sorry, I didn't make it explicit. My examples above are in the context of the customizer example you linked earlier. Here's a version that more clearly breaks out the python vs scad functionality and variable naming:

python:
```python
from solid2 import cube, text, translate, union, scad_render_to_file, scad_inline, \
                   CustomizerDropdownVariable, CustomizerSliderVariable, ScadVariable

#register all the custom variables you want to use
objects_var = CustomizerDropdownVariable("objects", 4, [2, 4, 6])
side_var = CustomizerSliderVariable("side", 4)

scene = union()

i_var = ScadVariable("i", 1)
one_var = ScadVariable("one", 1)
scene += scad_inline("for (")
scene += i_var
scene += scad_inline("=[")
scene += one_var
scene += scad_inline(":")
scene += objects_var
scene += scad_inline("]){")
scene += cube(side_var).translateX(2*i_var*side_var)
scene += scad_inline("}")

scad_render_to_file(scene)
```

openscad:
```openscad
objects = 4; //[2, 4, 6]
side = 4; //[]
i = 1;
one = 1;

union() {
	for (	i	=[	one	:	objects	]){	translate(v = [((2 * i) * side), 0, 0]) {
		cube(size = side);
	}
	}}
```

A "dumb" first draft of `scad_for` would do something like that, just outputting the appropriate scad source content around the parameters. In this case, I hand crafted the range expression.

PS: We'd probably also want to be able to serialize literals. `scene += 1` is currently an error because the python numeric type doesn't have a `_render()` method. My temporary workaround was to just put one in a variable.

PPS: also `ScadVariable`s should be broken up from `Customizer*Variable`s so they don't show up in Customizer. I might take a crack at that separately, or at least open an issue for it if I don't.

**Comment by jeff-dh** [comment]
What about this:

```python
def scad_for_test():
    def scad_range(start, stop, step=1):
        return f"[{start}:{step}:{stop}]"
    def scad_for(scadrange, loop_lambda):
        body = loop_lambda(ScadValue("scad_for_index"))._render()
        return scad_inline(f"for(scad_for_index = {scadrange})\n{{\n{indent(body)}\n}}\n")

    return scad_for(scad_range(0, 7), lambda i : cube(1).up(i * 2))
```

it seems to work

**Comment by jeff-dh** [comment]
```python
def scad_if_test():
    def scad_if(condition, if_body, else_body):
        c = f"if({condition})\n{{\n{if_body._render()}\n}}\nelse\n{{\n{else_body._render()}\n}}"
        return scad_inline(c)

    c = cube(2).left(5)
    return scad_if(get_animation_time() < 0.5, c, c.color("red"))
```

**Comment by jeff-dh** [comment]
fixed by [8fa8f64](https://github.com/jeff-dh/SolidPython/commit/8fa8f64aac96ff20c7c8b12b72c4affd240418d5)

```python
from solid2 import CustomizerDropdownVariable, \
                   CustomizerSliderVariable, \
                   cube, scad_range, scad_for, scad_if

number_of_cubes = CustomizerSliderVariable("number_of_cubes", 7)
alt_color = CustomizerDropdownVariable("alt_color", "red", ["red", "blue", "green"])

def f_loop(i):
    c = scad_if(i % 2 == 0, cube(), cube().color(alt_color))
    return c.up(i*2)

cubes_stacked = scad_for(scad_range(0, number_of_cubes-1), f_loop)

cubes_stacked.save_as_scad()
```

**Comment by sparr** [comment]
Thanks! I look forward to giving these a try.
